Sources/SecureXPC/Client/XPCClient.swift

```diff
@@ -157,87 +157,153 @@ public class XPCClient {
-        let encoded = try Request(route: route.route).dictionary
-        try sendWithReply(encoded: encoded, withReply: reply)
+                                   withResponse handler: @escaping XPCResponseHandler<R>) {
+        do {
+            let encoded = try Request(route: route.route).dictionary
+            sendWithResponse(encoded: encoded, withResponse: handler)
+        } catch {
+            handler(.failure(.encodingError(String(describing: error))))
+        }
     }
     
-    /// Sends a message and provides the reply as either a message on success or an error on failure.
+    /// Sends a message and provides the response as either a reply on success or an error on failure.
     ///
     /// - Parameters:
     ///    - message: Message to be sent.
     ///    - route: The server route which should handle this message.
-    ///    - withReply: A function or closure to receive the message's reply.
-    /// - Throws: If unable to encode the message or route or the server throws an error. No error will be thrown if communication with the server fails.
+    ///    - withResponse: A function or closure to receive the message's response.
     public func sendMessage<M: Encodable, R: Decodable>(_ message: M,
                                                         route: XPCRouteWithMessageWithReply<M, R>,
-                                                        withReply reply: @escaping XPCReplyHandler<R>) throws {
-        let encoded = try Request(route: route.route, payload: message).dictionary
-        try sendWithReply(encoded: encoded, withReply: reply)
+                                                        withResponse handler: @escaping XPCResponseHandler<R>) {
+        do {
+            let encoded = try Request(route: route.route, payload: message).dictionary
+            sendWithResponse(encoded: encoded, withResponse: handler)
+        } catch {
+            handler(.failure(.encodingError(String(describing: error))))
+        }
     }
     
-    /// Does the actual work of sending an XPC message which receives a reply.
-    private func sendWithReply<R: Decodable>(encoded: xpc_object_t,
-                                             withReply reply: @escaping XPCReplyHandler<R>) throws {
-        xpc_connection_send_message_with_reply(try getConnection(), encoded, nil, { xpcResponse in
+    /// Does the actual work of sending an XPC message which receives a response.
+    private func sendWithResponse<R: Decodable>(encoded: xpc_object_t,
+                                                withResponse handler: @escaping XPCResponseHandler<R>) {
+        // Get the connection or inform the handler of failure and return
+        var connection: xpc_connection_t?
+        do {
+            connection = try getConnection()
+        } catch {
+            if let error = error as? XPCError {
+                handler(.failure(error))
+            } else {
+                handler(.failure(.unknown))
```

Comment:
This kind of logic is repeated a few times, perhaps it would be nice to put it in an initializer on `XPCError`

Reply:
Heh, was thinking that exactly as I wrote this a few times, but didn't want to make the PR too big. Am planning to do this soon.